contrib/nydusify/pkg/backend/oss.go

```diff
@@ -134,102 +131,119 @@ func (b *OSSBackend) Upload(ctx context.Context, blobID, blobPath string, size i
-		}
+	var parts []oss.UploadPart
+	for p := range partsChan {
+		parts = append(parts, p)
+	}
 
-		close(partsChan)
+	ms := multipartStatus{
+		imur:          &imur,
+		parts:         parts,
+		blobObjectKey: blobObjectKey,
+		crc64:         crc64,
+		crc64ErrChan:  crc64ErrChan,
+	}
+	b.msMutex.Lock()
+	defer b.msMutex.Unlock()
+	b.ms = append(b.ms, ms)
 
-		var parts []oss.UploadPart
-		for p := range partsChan {
-			parts = append(parts, p)
-		}
+	logrus.Debugf("uploaded blob %s, costs %s", blobObjectKey, time.Since(start))
 
-		_, err = b.bucket.CompleteMultipartUpload(imur, parts)
-		if err != nil {
-			return nil, err
+	return &desc, nil
+}
+
+func (b *OSSBackend) Finalize(cancel bool) error {
+	b.msMutex.Lock()
+	defer b.msMutex.Unlock()
+
+	for _, ms := range b.ms {
+		if cancel {
+			// If there is any failure during conversion process, it will
+			// cause the uploaded blob to be left on oss, and these blobs
+			// are hard to be GC-ed, so we need always to use the multipart
+			// upload, and should call the `AbortMultipartUpload` method to
+			// prevent blob residue as much as possible once any error happens
+			// during conversion process.
+			if err := b.bucket.AbortMultipartUpload(*ms.imur); err != nil {
+				logrus.WithError(err).Warn("abort multipart upload")
+			} else {
+				logrus.Warnf("blob upload has been aborted: %s", ms.blobObjectKey)
+			}
+			continue
 		}
-	} else {
-		reader, err := os.Open(blobPath)
+
+		_, err := b.bucket.CompleteMultipartUpload(*ms.imur, ms.parts)
 		if err != nil {
-			return nil, err
+			return errors.Wrap(err, "complete multipart upload")
 		}
-		defer reader.Close()
-		err = b.bucket.PutObject(blobObjectKey, reader)
+
+		props, err := b.bucket.GetObjectDetailedMeta(ms.blobObjectKey)
 		if err != nil {
-			return nil, err
+			return errors.Wrapf(err, "get object meta")
 		}
-	}
 
-	props, err := b.bucket.GetObjectDetailedMeta(blobObjectKey)
-	if err != nil {
-		return nil, errors.Wrapf(err, "get object meta")
-	}
+		// Try to validate blob object integrity if any crc64 value is returned.
+		if value, ok := props[http.CanonicalHeaderKey("x-oss-hash-crc64ecma")]; ok {
+			if len(value) == 1 {
+				uploadedCrc, err := strconv.ParseUint(value[0], 10, 64)
+				if err != nil {
+					return errors.Wrapf(err, "parse uploaded crc64")
+				}
 
-	// Try to validate blob object integrity if any crc64 value is returned.
-	if value, ok := props[http.CanonicalHeaderKey("x-oss-hash-crc64ecma")]; ok {
-		if len(value) == 1 {
-			uploadedCrc, err := strconv.ParseUint(value[0], 10, 64)
-			if err != nil {
-				return nil, errors.Wrapf(err, "parse uploaded crc64")
-			}
+				err = <-ms.crc64ErrChan
+				if err != nil {
+					return errors.Wrapf(err, "calculate crc64")
+				}
 
-			err = <-crc64ErrChan
-			if err != nil {
-				return nil, errors.Wrapf(err, "calculate crc64")
-			}
+				if uploadedCrc != ms.crc64 {
+					return errors.Errorf("crc64 mismatch, uploaded=%d, expected=%d", uploadedCrc, ms.crc64)
+				}
```

Comment:
Maybe not related to this PR, but what happens if crc64 does not match? Do we clean up the OSS object?

Reply:
We can only get the crc64 property of the OSS object after the upload is complete, if crc64 doesn't match, we can't actually do anything (we can't delete the object), only can return an error and stop pushing the image manifest.

In fact, I think the crc64 check is no longer needed, we are long-term use in production has not shown crc64 mismatches, so keeping it would only consume resources. We can add a UT to ensure that there is no data corruption with concurrent multipart uploads.